apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/HiPerfPublisher.cs

```diff
@@ -0,0 +1,626 @@
+            /// </summary>
+            public long? BatchMaxByteCount { get; set; }
+
+            internal void Validate()
+            {
+                if (Clients != null)
+                {
+                    GaxPreconditions.CheckArgument(Clients.Count > 0, nameof(Clients), "If non-null, then must contain at least one element");
+                    GaxPreconditions.CheckArgument(Clients.All(x => x != null), nameof(Clients), "If non-null, then all elements must be non-null");
+                }
+                // Fairly arbitrary upper limit.
+                GaxPreconditions.CheckArgumentRange(ClientCount ?? 1, nameof(ClientCount), 1, 256);
+                if (BatchingSettings != null)
+                {
+                    GaxPreconditions2.CheckArgumentRange(BatchingSettings.ElementCountThreshold ?? 1, nameof(BatchingSettings.ElementCountThreshold), 1, ApiMaxRequestElementCount);
+                    GaxPreconditions2.CheckArgumentRange(BatchingSettings.RequestByteThreshold ?? 1, nameof(BatchingSettings.RequestByteThreshold), 1, ApiMaxRequestByteCount);
+                    GaxPreconditions.CheckArgument((BatchingSettings.DelayThreshold ?? TimeSpan.FromSeconds(1)) > TimeSpan.Zero, nameof(BatchingSettings.DelayThreshold), "Must be positive");
+                }
+                GaxPreconditions2.CheckArgumentRange(BatchMaxElementCount ?? 1, nameof(BatchMaxElementCount), 1, ApiMaxRequestElementCount);
+                GaxPreconditions2.CheckArgumentRange(BatchMaxByteCount ?? 1, nameof(BatchMaxByteCount), 1, ApiMaxRequestByteCount);
+            }
+
+            /// <summary>
+            /// Create a deep clone of this object.
+            /// </summary>
+            /// <returns>A deep clone of this object.</returns>
+            public Settings Clone() => new Settings(this);
+        }
+
+        /// <summary>
+        /// A snapshot of the flow state of a <see cref="HiPerfPublisher"/>.
+        /// </summary>
+        public struct FlowState
+        {
+            /// <summary>
+            /// Instantiate a <see cref="FlowState"/>.
+            /// </summary>
+            /// <param name="elementCount">The number of elements (messages) currently queued.</param>
+            /// <param name="byteCount">The number of bytes currently queued.</param>
+            public FlowState(long elementCount, long byteCount)
+            {
+                ElementCount = elementCount;
+                ByteCount = byteCount;
+            }
+
+            /// <summary>
+            /// The number of elements (messages) currently queued.
+            /// </summary>
+            public long ElementCount { get; }
+
+            /// <summary>
+            /// The number of bytes currently queued.
+            /// </summary>
+            public long ByteCount { get; }
+        }
+
+        // All defaults taken from Java (reference) implementation.
+
+        /// <summary>
```

Comment:
See above comment about `BatchingSettings` thresholds.
Yes, I agree that BatchingSettings would probably be better immutable, and the exact structure of this code could change.
But `BatchingSettings` includes a delay, which isn't relevant here. I dislike have irrelevant settings available.